python/pyspark/sql/connect/expressions.py

```diff
@@ -89,7 +90,16 @@ class Expression:
     """
 
     def __init__(self) -> None:
-        pass
+        origin = current_origin()
+        fragment = origin.fragment
+        call_site = origin.call_site
```

Comment:
I think this is what I don't like about this approach at least for Spark Connect. There is no need to use the thread-local for this propagation. 

The outside wrapper manipulates the global state for the constructor to pick up here. Shouldn't we simply generate the origin here?

Reply:
I don't quite like this too actually. Yes, we can pass the explicit callsite around. It happened to be like this to match the implementation across Catalyst Optimizer, Scala, and PySpark w/ Spark Classic.

Let me revisit this to fix up Spark Classic side together and see if we can do.